ebean-core/src/test/java/org/tests/update/TestSqlUpdateBatch.java

```diff
@@ -0,0 +1,67 @@
+package org.tests.update;
+
+import io.ebean.BaseTestCase;
+import io.ebean.DB;
+import io.ebean.SqlUpdate;
+import io.ebean.Transaction;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Testclass that ensures the correct behaviour of SqlUpdate in combination with batching.
+ *
+ * @author Jonas P&ouml;hler, FOCONIS AG
+ */
+public class TestSqlUpdateBatch extends BaseTestCase {
+
+  /**
+   * This test ensures the correct behaviour of update batching in the {@code BatchedPstmtHolder} after a flush. Two batch updates
+   * are executed with each a batchsize of 21, which is the default limit for automatic flushing on {@code addBatch()}. After that
+   * the statements are each executed with {@code executeBatch()}. The desired behaviour is, that each execution succeeds although
+   * technically the batch was already flushed.
+   */
+  @Test
+  public void testTwoParallelBatches() {
+    try (Transaction txn = DB.beginTransaction()) {
+      // Dummy updates, that effectively do nothing, but ebean doesn't need to know this.
+      final SqlUpdate update = DB.sqlUpdate("update uuone set name = ? where 0=1");
+      final SqlUpdate delete = DB.sqlUpdate("delete from uuone where ?=-1");
+      // txn.setBatchSize(40);
+
+      for (int i = 0; i <= 20; i++) {
+        update
+          .setParameter(1, String.valueOf(i))
+          .addBatch();
+        delete
+          .setParameter(1, String.valueOf(i))
+          .addBatch();
+      }
+
+      delete.executeBatch();
+      update.executeBatch();
+      txn.commit();
+    }
+  }
+
+  /**
+   * This test checks that for a batch update a correct array with update counts is returned. If a update with 40 entries is
+   * executed, it is expected, that {@code executeBatch()} returns a array with 40 elements each containing the update count for
+   * the given parameters.
+   */
+  @Test
+  public void testBatchReturnArrayLength() {
+    try (Transaction txn = DB.beginTransaction()) {
+      txn.setBatchSize(100); // something bigger than 40
+      // Dummy update, that effectively does nothing, but ebean doesn't need to know this.
+      final SqlUpdate update = DB.sqlUpdate("update uuone set name = ? where 0=1");
+      for (int i = 0; i < 40; i++) {
+        update
+          .setParameter(1, String.valueOf(i))
+          .addBatch();
+      }
+      assertEquals(40, update.executeBatch().length);
```

Comment:
Hello Rob, our intention was, that the returned array of `update.executeBatch()` always matches the amount of `addBatch()` calls and does not depend on batch size settings.
 In most cases, we are not really interested in the row count, so it would be OK if `executeBatch` returns `void`
What do you think of a callback

```java
interface BatchCallback {
     rowsUpdated(int[] rowCounts)
}

final SqlUpdate update = DB.sqlUpdate("update uuone set name = ? where 0=1");
List<Integer> allCounts = new ArrayList;
update.setBatchExecutedCallback(rowCounts => allCounts.addAll(Arrays.asList(rowCounts)));
// do X addBatches
// allCounts should always contain X elements

```

Reply:
>  update.executeBatch() always matches the amount of addBatch() calls and does not depend on batch size settings.

Well yes and I think that would be nice.  The issue with making that change is that the other bug/issue is totally dominated my "brain space" in the sense it is the bigger / tougher issue that I wanted to sort out first.  

We can look at this `executeBatch()` `addBatch` properly later like maybe in a week but in my mind that means it the suggested alternative is that `executeBatch()` does actually execute the batch (and that the batch isn't auto flushed earlier).  

> BatchCallback

My gut says no on that and says ... lets make executeBatch() actually execute the batch and stop the early flush that occurs (get it to ignore transaction batch size).

Reply:
So approximately ... addBatch() needs to explicitly go deeper into the internals such that it "stops automatic batch execute on transaction batch size" from happening (and then see if that change is appropriate / ok).